### PR TITLE
WIP - Update Shipment with Original dates on BeforeSave()

### PR DESCRIPTION
**Pull request by chrisgilmerproj**
## Description

It appears that some shipments in staging are missing dates.  This code change wants to make the dates less fragile by ensuring they exist on the shipment each time it is saved (but only if they are missing).  We figured this out during a bug found while testing staging.

## Reviewer Notes

- [x] Refactor calculateMoveDates functions so they can be used in the BeforeSave() function
- [x] Calculate dates in BeforeSave() method

## Setup

Run through the service flow and create an HHG shipment.  Then go in the DB and remove the OriginalPackDate or OriginalDeliveryDate from that shipment.  Now do any updating on the shipment via the TSP app that might call PATCH or POST methods.  You should see those dates update in the DB.

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](./docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](./docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](./docs/database.md#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
* [ ] There are no aXe warnings for UI.
* [ ] This works in IE.
* Any new client dependencies (Google Analytics, hosted libraries, CDNs, etc) have been:
  * [ ] Communicated to @willowbl00
  * [ ] Added to the list of [network dependencies](https://github.com/transcom/mymove#client-network-dependencies)
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/161495172) for this change

### CODE REVIEW

File: pkg/models/shipment.go [hunk 1]
```diff
@@ -224,6 +224,21 @@ func (s *Shipment) BeforeSave(tx *pop.Connection) error {
 		s.TrafficDistributionList = trafficDistributionList
 	}
 
+	if s.RequestedPickupDate && (s.OriginalPackDate == nil || s.OriginalDeliveryDate == nil) {
+		summary, _ := CalculateMoveDatesFromShipment(s)
```
Comment (chrisgilmerproj) [inline pkg/models/shipment.go]: It's unclear if I should be using this method or `CalculateMoveDates` and the differences between the two.  My feeling is that `CalculateMoveDates` cares a fair amount about the transit distance.  However, that seems to get lost when doing a calculation based just from the shipment.  Is there any reason these shouldn't be combined into one method? Or one method that takes a `*models.Move` and one that takes `*models.Shipment` that fall back to the same underlying logic?

**Comment by chrisgilmerproj** [comment]
Also, as a reference, should be making sure I don't conflict with @kimallen 's branch https://github.com/transcom/mymove/compare/kma-160866959-change-sm-hhg-homepage-by-status

File: pkg/models/shipment.go (hunk 1 above)
Comment (reggieriser) [reply pkg/models/shipment.go]: Are you sure that the record on staging that you're trying to fix has non-nil EstimatedPackDays and EstimatedTransitDays on it?  It appears that those are required for this calculateMoveDatesFromShipment to work.  @kimallen may be the better person to talk about the differences between the two methods -- this additional way of calculating move dates came out of one of her stories I believe.

Comment (chrisgilmerproj) [reply pkg/models/shipment.go]: I'm unsure! But I'd rather be robust.  So I'm thinking really I ought to use `CalculateMoveDates` instead.  I think that would be the better way to do it as long as its deterministic.

Comment (kimallen) [reply pkg/models/shipment.go]: I have changed calculateMoveDatesFromShipment on my branch such that it no longer uses the `EstimatedPackDays` or the `EstimatedTransitDays` but uses the `OriginalPackDate` and `OriginalDeliveryDate` on the assumption that those values will have been saved already.

File: pkg/models/shipment.go [hunk 2]
```diff
@@ -224,6 +224,22 @@ func (s *Shipment) BeforeSave(tx *pop.Connection) error {
 		s.TrafficDistributionList = trafficDistributionList
 	}
 
+	// Ensure that OriginalPackDate and OriginalDeliveryDate are set if we know RequestedPickupDate,
+	// EstimatedPackDays and EstimatedTransitDays.
```
Comment (chrisgilmerproj) [inline pkg/models/shipment.go]: This continues to be the main thing I want to accomplish with this PR.  However, I feel like I'm getting further from where I want to be with each change.

**Comment by chrisgilmerproj** [comment]
Closing and following up in https://github.com/transcom/mymove/pull/1283

**Pull request closed without merging**